forem/forem
# Admin Save Confirmation - Removing Badge Achievements

## What type of PR is this? (check all applicable)


## Description
In this PR, I refactor the process of deleting Badge Achievements, by presenting a Confirmation Modal (utilizing the Preact `<Modal />`) before completing the destructive action. This flow is called "Heavy Confirmation" in the [Save-Confirmation RFC](https://github.com/forem/rfcs/blob/237480c4b915a64fc17df8f525b0a429185a1605/text/0078-admin-save-confirmation.md).
I also create a `CustomEvent` that generates error alerts in response to unsuccessful async requests. These error alerts are to replace the server-side `flash[:error]` notices that Rails generates.


### IMPORTANT NOTES
- The RFC calls for the refactoring of over a dozen destructive-action types in Admin. This current PR implements heavy-confirmation for removing Badge Achievements only.
- I want to get the flow correct in this present PR, before I expand implementation to the other actions.
- I will implement heavy-confirmation flows for the other destructive actions in separate PRs.
- I choose this approach so that my PRs do not balloon in size. The unrefactored destructive actions can continue using the native `data: confirm` flow successfully, until I get to them.

---


## QA Instructions, Screenshots, Recordings

1. Login to `http://localhost:xxxx/admin/` as a `super_admin`. Navigate to `Content Manager`, then `Badges`, then click on the `Achievements` tab.
2. Click the `Remove` button for one of the achievements shown. A modal titled `Confirm changes` should popup.
3. Click either the `Discard Changes` or `X` button at the top right corner of the modal. The modal should close without any actions being taken.
4. Click any `Remove` button again. Type in some random text, then click `Confirm changes`. A yellow warning notice should appear, telling you that your text does not match the expected entry.
5. Now type in the correct confirmation text and click `Confirm Changes`. A confirmatory Snackbar should appear in the bottom-left, and the associated badge achievement gone from the list.
6. The Cypress spec I added tests the error alerts for the unhappy path.

### UI accessibility concerns?
All UI accessibility concerns have been address through PR review.

## Added/updated tests?

      have not been included_


## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will update the Admin Guide
      replace this line with details on why this change doesn't need to be
      shared_

## What gif best describes this PR or how it makes you feel?
![Olivia Benson speaking](https://i.giphy.com/media/mA13FdrkYyZTrEfPCf/giphy.webp)
